src/librustc_mir/interpret/eval_context.rs

```diff
@@ -546,57 +566,143 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tc
         }
     }
 
-    pub(super) fn pop_stack_frame(&mut self) -> EvalResult<'tcx> {
+
+    pub(super) fn pop_stack_frame_internal(&mut self) -> EvalResult<'tcx, StackPopCleanup> {
+        use UnwindingState::*;
         info!("LEAVING({}) {}", self.cur_frame(), self.frame().instance);
         ::log_settings::settings().indentation -= 1;
         let frame = self.stack.pop().expect(
             "tried to pop a stack frame, but there were none",
         );
-        M::stack_pop(self, frame.extra)?;
+        let stack_pop_info = M::stack_pop(self, frame.extra)?;
+        let prev = self.unwinding;
```

Comment:
`prev` is not very descriptive... maybe `prev_unwinding`?

Also I am still trying to figure out what this state machine does and what it corresponds to in a real execution.

Reply:
There are three possible states that we can be in:
1. Normal execution
2. Stack unwinding
3. Cleanup block during stack unwinding

I'm not sure if `libpanic_unwind` explicitly does this kind of validation, as it relies on platform-specific unwinding mechanisms for part of the work. This `match` is mainly to catch errors in Miri or the MIR itself.

Reply:
I remain doubtful about doing this kind of state tracking *at all*, but I am *very* doubtful about distinguishing (2) and (3) in your list. Do I understand correctly that this is about distinguishing whether we are inside a destructor or whether we are inside the glue code that calls all destructors?